<a href="https://colab.research.google.com/github/salianbharat/BE_DS_Assignment/blob/main/Naive_Bayes__Mashroom.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Naive Bayes Model for Mashroom Dataset

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
# For preprocessing data, including one-hot encoding, label encoding, and binning
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, KBinsDiscretizer

# For splitting the dataset into training and testing sets
from sklearn.model_selection import train_test_split

# For building the Naive Bayes model
from sklearn.naive_bayes import GaussianNB

# For evaluating the model's performance
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [3]:
mashroom = pd.read_csv('/content/drive/MyDrive/New Assignment/mushroom_cleaned.csv')

In [4]:
mashroom.head()

cap-diameter  cap-shape  gill-attachment  gill-color  stem-height  \
0          1372          2                2          10     3.807467   
1          1461          2                2          10     3.807467   
2          1371          2                2          10     3.612496   
3          1261          6                2          10     3.787572   
4          1305          6                2          10     3.711971   

   stem-width  stem-color    season  class  
0        1545          11  1.804273      1  
1        1557          11  1.804273      1  
2        1566          11  1.804273      1  
3        1566          11  1.804273      1  
4        1464          11  0.943195      1

In [5]:
mashroom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54035 entries, 0 to 54034
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   cap-diameter     54035 non-null  int64  
 1   cap-shape        54035 non-null  int64  
 2   gill-attachment  54035 non-null  int64  
 3   gill-color       54035 non-null  int64  
 4   stem-height      54035 non-null  float64
 5   stem-width       54035 non-null  int64  
 6   stem-color       54035 non-null  int64  
 7   season           54035 non-null  float64
 8   class            54035 non-null  int64  
dtypes: float64(2), int64(7)
memory usage: 3.7 MB


In [6]:
mashroom.shape

(54035, 9)

In [7]:
mashroom.describe()

cap-diameter     cap-shape  gill-attachment    gill-color  \
count  54035.000000  54035.000000     54035.000000  54035.000000   
mean     567.257204      4.000315         2.142056      7.329509   
std      359.883763      2.160505         2.228821      3.200266   
min        0.000000      0.000000         0.000000      0.000000   
25%      289.000000      2.000000         0.000000      5.000000   
50%      525.000000      5.000000         1.000000      8.000000   
75%      781.000000      6.000000         4.000000     10.000000   
max     1891.000000      6.000000         6.000000     11.000000   

        stem-height    stem-width    stem-color        season         class  
count  54035.000000  54035.000000  54035.000000  54035.000000  54035.000000  
mean       0.759110   1051.081299      8.418062      0.952163      0.549181  
std        0.650969    782.056076      3.262078      0.305594      0.497580  
min        0.000426      0.000000      0.000000      0.027372      0.000000  
25%        0.270997    421.000000      6.000000      0.888450      0.000000  
50%        0.593295    923.000000     11.000000      0.943195      1.000000  
75%        1.054858   1523.000000     11.000000      0.943195      1.000000  
max        3.835320   3569.000000     12.000000      1.804273      1.000000

**Observations from Data Exploration:**
Dataset has 54035 rows and 9 columns.
Data type of all columns is numerical (float or integer).
All values are non-null.

In [8]:
mashroom['class'].unique()

array([1, 0])

Since Data is clean we will directly jump into preprocessing the data.

In [9]:
# Separate features and target variable
X = mashroom.drop('class', axis=1)
y = mashroom['class']

In [10]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [11]:
# Step 3: Train the Naive Bayes Model
# Initialize the Gaussian Naive Bayes model
gnb = GaussianNB()

In [12]:
# Fit the model to the training data
gnb.fit(X_train, y_train)

GaussianNB()

In [13]:
# Step 4: Evaluate the Model
# Predict the labels of the test set
y_pred = gnb.predict(X_test)

In [14]:
# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

Accuracy: 0.64


In [15]:
# Generate confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

Confusion Matrix:
[[3917 3432]
 [2454 6408]]


In [16]:

# Generate classification report
class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Classification Report:
              precision    recall  f1-score   support

           0       0.61      0.53      0.57      7349
           1       0.65      0.72      0.69      8862

    accuracy                           0.64     16211
   macro avg       0.63      0.63      0.63     16211
weighted avg       0.63      0.64      0.63     16211



In [17]:
# Step 5: Conclusion
print("The Naive Bayes classifier achieved an accuracy of {:.2f}. The classification report and confusion matrix provide detailed insights into the model's performance.".format(accuracy))

The Naive Bayes classifier achieved an accuracy of 0.64. The classification report and confusion matrix provide detailed insights into the model's performance.


## **Conclusion**

**Moderate Performance:**

The classifier performs moderately well, with an overall accuracy of 64%.
The precision and recall scores indicate that the classifier is better at identifying poisonous mushrooms than edible ones.

**Class Imbalance:**

There is a slight class imbalance with more poisonous mushrooms (8862) than edible ones (7349) in the test set.
The classifier's performance (higher recall and F1-score) is better for poisonous mushrooms, which might be due to the class imbalance.

Hence there is scope of optimization of the dataset- one way is to balance the dataset and then try it out.